lib/ui/painting.dart

```diff
@@ -4935,26 +4591,1632 @@ class PictureRecorder extends NativeFieldWrapperClass2 {
+    if (width <= 0 || height <= 0)
+      throw Exception('Invalid image dimensions.');
+    return _futurize(
+          (_Callback<Image?> callback) => _toImage(width, height, (_Image? image) {
+        if (image == null) {
+          callback(null);
+        } else {
+          callback(Image._(image));
+        }
+      }),
+    );
+  }
+
+  String? _toImage(int width, int height, _Callback<_Image?> callback) native 'Picture_toImage';
+
+  Rect _getBounds() {
+    final Float32List rect = _getBoundsF32();
+    return Rect.fromLTRB(rect[0], rect[1], rect[2], rect[3]);
+  }
+  Float32List _getBoundsF32() native 'Picture_getBounds';
+
+  @override
+  void dispose() native 'Picture_dispose';
+
+  @override
+  int get approximateBytesUsed native 'Picture_GetAllocationSize';
+}
+
+class _SkiaPictureRecorder extends NativeFieldWrapperClass2 implements PictureRecorder {
+  @pragma('vm:entry-point')
+  _SkiaPictureRecorder() { _constructor(); }
+  void _constructor() native 'PictureRecorder_constructor';
+
+  @override
+  Canvas _makeCanvas(Rect? cullRect) => _SkiaCanvas(this, cullRect);
+
+  @override
+  bool get isRecording => _canvas != null;
+
+  @override
+  Picture endRecording() {
+    if (_canvas == null)
+      throw StateError('PictureRecorder did not start recording.');
+    final Picture picture = _SkiaPicture._();
+    _endRecording(picture);
+    _canvas!._recorder = null;
+    _canvas = null;
+    return picture;
+  }
+
+  void _endRecording(Picture outPicture) native 'PictureRecorder_endRecording';
+
+  _SkiaCanvas? _canvas;
+}
+
+enum _CanvasOp {
+  setAA,
+  clearAA,
+  setDither,
+  clearDither,
+  setInvertColors,
+  clearInvertColors,
+  setFillStyle,
+  setStrokeStyle,
+
+  setCapsButt,
+  setCapsRound,
+  setCapsSquare,
+  setJoinsBevel,
+  setJoinsMiter,
+  setJoinsRound,
+
+  setStrokeWidth,
+  setMiterLimit,
+
+  setFilterQualityNearest,
+  setFilterQualityLinear,
+  setFilterQualityMipmap,
+  setFilterQualityCubic,
+
+  setColor,
+  setBlendMode,
+
+  setShader,
+  clearShader,
+  setColorFilter,
+  clearColorFilter,
+  setImageFilter,
+  clearImageFilter,
+
+  clearMaskFilter,
+  setMaskFilterNormal,
+  setMaskFilterSolid,
+  setMaskFilterOuter,
+  setMaskFilterInner,
+
+  save,
+  saveLayer,
+  saveLayerBounds,
+  restore,
+
+  translate,
+  scale,
+  rotate,
+  skew,
+  transform2x3,
+  transform3x3,
+
+  clipRect,
+  clipRectAA,
+  clipRectDiff,
+  clipRectAADiff,
+  clipRRect,
+  clipRRectAA,
+  clipPath,
+  clipPathAA,
+
+  drawPaint,
+  drawColor,
+
+  drawLine,
+  drawRect,
+  drawOval,
+  drawCircle,
+  drawRRect,
+  drawDRRect,
+  drawArc,
+  drawArcCenter,
+  drawPath,
+
+  drawPoints,
+  drawLines,
+  drawPolygon,
+  drawVertices,
+
+  drawImage,
+  drawImageRect,
+  drawImageNine,
+  drawAtlas,
+  drawAtlasColored,
+  drawAtlasCulled,
+  drawAtlasColoredCulled,
+
+  drawSkPicture,
+  drawDisplayList,
+  drawShadow,
+  drawShadowOccluded,
+}
+
+class _MatrixTransform {
```

Comment:
At the very least, it's probably worth documenting that these are utility methods for transforming affine matrices, and not meant to handle 4x4 matrix manipulation directly like vector_math.

Reply:
I've been thinking about this. I think `package:vector_math` is not the best vector math library for Flutter. Because all data structures are mutable there's a lot of defensive copying going on (how many instances of `Matrix4.identity` do we need?). It's also too pessimistic, always assuming the most general matrices in the system. In practice, however, most matrices flowing through our system are either identity or plain 2D matrices expressible using 6 values or less. Separating the API into mutable aggregators/builders and deeply immutable values we can optimize by reducing memory footprint and specializations (e.g. make multiplication by identity a noop, and simplify 2D-only multiplications).

I think the best approach would be to have `dart:ui` provide a matrix library optimized for Flutter's use-case and drop our dependency on `package:vector_math`. That's a big project, however, with breaking changes. It would be best to make that effort independent from this particular PR. So I think it's OK to have another copy here for now.

Reply:
My "soon" plan is to delete this code from Dart and instead compute bounds lazily. This would simplify a lot of the code here in producing the DL captures and we run through the data natively anyway to extract the Skia objects so we can compute bounds while we do that (or even more lazily).